File: srpc/packet.go
package srpc

// CloseHandler handles the stream closing with an optional error.
type CloseHandler = func(closeErr error)

// PacketHandler handles a packet.
//
// pkt is optional (can be nil)
// if closeErr is set, the stream is closed after pkt.
type PacketHandler = func(pkt *Packet) error

// PacketDataHandler handles a packet before it is parsed.
type PacketDataHandler = func(data []byte) error

// NewPacketDataHandler wraps a PacketHandler with a decoding step.
func NewPacketDataHandler(handler PacketHandler) PacketDataHandler {
	return func(data []byte) error {
		pkt := &Packet{}
		if err := pkt.UnmarshalVT(data); err != nil {
			return err
		}
		return handler(pkt)
	}
}

// Validate performs cursory validation of the packet.
func (p *Packet) Validate() error {
	switch b := p.GetBody().(type) {
	case *Packet_CallStart:
		return b.CallStart.Validate()
	case *Packet_CallData:
		return b.CallData.Validate()
	case *Packet_CallCancel:
		return nil
	default:
		return ErrUnrecognizedPacket
	}
}

// NewCallStartPacket constructs a new CallStart packet.
func NewCallStartPacket(service, method string, data []byte, dataIsZero bool) *Packet {
	return &Packet{Body: &Packet_CallStart{
		CallStart: &CallStart{
			RpcService: service,
			RpcMethod:  method,
			Data:       data,
			DataIsZero: dataIsZero,
		},
	}}
}

// Validate performs cursory validation of the packet.
func (p *CallStart) Validate() error {
	method := p.GetRpcMethod()
	if len(method) == 0 {
		return ErrEmptyMethodID
	}
	service := p.GetRpcService()
	if len(service) == 0 {
		return ErrEmptyServiceID
	}
	return nil
}

// NewCallDataPacket constructs a new CallData packet.
func NewCallDataPacket(data []byte, dataIsZero bool, complete bool, err error) *Packet {
	var errStr string
	if err != nil {
		errStr = err.Error()
	}
	return &Packet{Body: &Packet_CallData{
		CallData: &CallData{
			Data:       data,
			DataIsZero: dataIsZero,
			Complete:   err != nil || complete,
			Error:      errStr,
		},
	}}
}

// NewCallCancelPacket constructs a new CallCancel packet with cancel.
func NewCallCancelPacket() *Packet {
	return &Packet{Body: &Packet_CallCancel{CallCancel: true}}
}

// Validate performs cursory validation of the packet.
func (p *CallData) Validate() error {
	if len(p.GetData()) == 0 && !p.GetComplete() && len(p.GetError()) == 0 && !p.GetDataIsZero() {
		return ErrEmptyPacket
	}
	return nil
}

File: srpc/packet.cpp
//go:build deps_only

#include "packet.hpp"

#include "rpcproto.pb.h"

namespace starpc {

PacketDataHandler NewPacketDataHandler(PacketHandler handler) {
  return [handler](const std::string& data) -> Error {
    srpc::Packet pkt;
    if (!pkt.ParseFromString(data)) {
      return Error::InvalidMessage;
    }
    return handler(pkt);
  };
}

Error ValidatePacket(const srpc::Packet& pkt) {
  switch (pkt.body_case()) {
    case srpc::Packet::kCallStart:
      return ValidateCallStart(pkt.call_start());
    case srpc::Packet::kCallData:
      return ValidateCallData(pkt.call_data());
    case srpc::Packet::kCallCancel:
      return Error::OK;
    default:
      return Error::UnrecognizedPacket;
  }
}

Error ValidateCallStart(const srpc::CallStart& pkt) {
  if (pkt.rpc_method().empty()) {
    return Error::EmptyMethodID;
  }
  if (pkt.rpc_service().empty()) {
    return Error::EmptyServiceID;
  }
  return Error::OK;
}

Error ValidateCallData(const srpc::CallData& pkt) {
  if (pkt.data().empty() && !pkt.complete() && pkt.error().empty() && !pkt.data_is_zero()) {
    return Error::EmptyPacket;
  }
  return Error::OK;
}

std::unique_ptr<srpc::Packet> NewCallStartPacket(
    const std::string& service,
    const std::string& method,
    const std::string& data,
    bool data_is_zero) {
  auto pkt = std::make_unique<srpc::Packet>();
  auto* call_start = pkt->mutable_call_start();
  call_start->set_rpc_service(service);
  call_start->set_rpc_method(method);
  call_start->set_data(data);
  call_start->set_data_is_zero(data_is_zero);
  return pkt;
}

std::unique_ptr<srpc::Packet> NewCallDataPacket(
    const std::string& data,
    bool data_is_zero,
    bool complete,
    Error err) {
  auto pkt = std::make_unique<srpc::Packet>();
  auto* call_data = pkt->mutable_call_data();
  call_data->set_data(data);
  call_data->set_data_is_zero(data_is_zero);
  call_data->set_complete(err != Error::OK || complete);
  if (err != Error::OK) {
    call_data->set_error(ErrorString(err));
  }
  return pkt;
}

std::unique_ptr<srpc::Packet> NewCallCancelPacket() {
  auto pkt = std::make_unique<srpc::Packet>();
  pkt->set_call_cancel(true);
  return pkt;
}

}  // namespace starpc
